### PR TITLE
Feature/add travis ci

### LINKED ISSUE #174: install_bootstrap errors when extra newline characters are entered
When answering the questions in install_bootstrap, if the user quickly presses ENTER or RETURN in between questions or when bringing up containers, extra newline characters are added to the ```redwood_access_token``` and/or ```storage_access_token``` environment variables in the ```boardwalk.config``` and ```action.config``` files (respectively).
When install_bootstrap is run a second time, this causes  the following commands...

```source <(jq -r "to_entries|map(\"\(.key)=\(.value|tostring)\")|.[]" boardwalk_launcher_config/boardwalk.config)```

and ```source <(jq -r "to_entries|map(\"\(.key)=\(.value|tostring)\")|.[]" action_launcher_config/action.config)``` 

...to result in an error, with none of the environment variables being properly set.

Possible cause: Using Git Bash terminal? Haven't tried reproducing using any other terminal.

### PR DESCRIPTION
**Pull request by jnkrupp**
Adds basic TravisCI testing to dcc-ops
* Executing the bootstrap script with the flag `-t` allows it to be run by taking in encrypted environment variables from the TravisCI repository settings
* Currently just runs `install_bootstrap`, deploying the Docker images and testing the script itself

Adds a second flag `-f` to `install_bootstrap` so it can be run using only default values from the `.config` files
* Doesn't require any user input
* Runs without using external Redwood databases
* Runs without recovering metadata from a backup

Also fixes #174 by adding a sed command to `install_bootstrap` that cleans the `.config` files, removing any ^M characters and/or extra newlines which cause parsing to fail

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+   - ELASTIC_SEARCH_PORT="9200"
+   - TOUCH_FILE_DIRECTORY="travis-cgp-action-service"
+
+before_install:
+ - sudo apt-get -y update
+# clone the github repo and cd into it
+ - git clone https://github.com/BD2KGenomics/dcc-ops.git
+ - cd dcc-ops
+ - git checkout feature/add-travis-ci
+
```
Comment (mikebaumann) [inline .travis.yml]: Are you sure you need to clone the repo and checkout this branch?
I've not seen the need to do this in other .travis.yml files.
In any case, it won't work after this PR (on `feature/add-travis-ci` is merged and the current branch is deleted.
Travis should be run against whatever the current branch is (for a given commit or PR).

File: install_bootstrap [hunk 2]
```diff
@@ -55,24 +55,69 @@ function ask_question()
...
   eval "${!4}=\"$answer\""
 }
 
+
+# Clean a file and remove all instances of ^M (carriage return) adjacent to a newline character.
+# Sometimes, when the user presses ENTER/RETURN, a carriage return character can get stuck in one of
+# the config files. This function was created to prevent errors when reading environment variables from config files.
+#   Arguments:
+#     $1 - file to remove carriage return characters from
+function remove_empty_lines {
```
Comment (mikebaumann) [inline install_bootstrap]: This whole function seems unnecessarily complex.
Blank lines are often deleted simply using `sed ‘/^$/d’` as described here:
https://www.cyberciti.biz/faq/using-sed-to-delete-empty-lines/

Is there a reason that won't work here?

File: install_bootstrap [hunk 3]
```diff
@@ -150,6 +195,8 @@ while [[ "${install_commons_launcher^^}" != 'Y' &&  "${install_commons_launcher^
   fi
 done
 
+
```
Comment (mikebaumann) [inline install_bootstrap:167]: Too many blank lines.

File: install_bootstrap [hunk 4]
```diff
@@ -173,222 +220,299 @@ fi
...
+            ;;
+      # Check for file input instead of whether or not the user would like to run consonance_launcher
+      # (We need only one environment variable from this, PEM_PATH, which is used after the mustache commands below)
+      file) # Read the config file if it exists
+            if [ -f commons_launcher_config/consonance.config ] ; then
+              source <(jq -r "to_entries|map(\"\(.key)=\(.value|tostring)\")|.[]" commons_launcher_config/consonance.config)
+            fi
+            ;;
+      # For testing outside of TravisCI servers only: Create environment variables from the config file instead of using the ones on the TravisCI console
+#      travis) # add environment variables to various files
```
Comment (mikebaumann) [inline install_bootstrap]: Which line does the comment on line 242 apply to, 237 or 243?
If 243, there seems to be an inconsistency, as the .travis.yml invokes `install_bootstrap` with the `-t`, which means `travis` mode, but line 242 states "Fort testing outside of TravisCI servers only".
In any case, it should be more clear what line 242 applies to.

**Review by mikebaumann — CHANGES_REQUESTED** [review]
In addition to comments made in the changed code, here are some general comments/concerns:

How are the AWS keys being handled securely?
This is not clear to me from looking at these changes.
It seems one potentially good way of handling them is as described in the TravisCI documentation, here:
https://docs.travis-ci.com/user/encryption-keys/
However, that does not seem to be used.
How do you know the keys you are using, however you are doing it, are actually working?
The Docker images are deploying, yet there are no tests that the services are actually functional.

This is good progress, yet there need to be additional testing to verify the deployed system is actually functioning correctly. At a minimum, to complete this PR, I think the log files produced by each of the services should be automatically checked for errors/exceptions. If it is the case that some errors/exceptions are logged that are "expected" and don't actually indicate a problem, then support should be added for an "ignore" list that contains, for example, regular expressions of the errors/exceptions that can safely be ignored.

Beyond this, perhaps in another PR, a good next step would be load a file/bundle into Redwood using Spinnaker then verifying that the file/bundle is available from the dashboard service.

There should be updates to the README.md describing:
- The new options added to `install_bootstrap` and how to use them.
- How to setup `dcc-ops` for TravisCI (including buckets that need to be created, etc.)

Also, please make your commit messages descriptive!
For example, please read and apply:
https://chris.beams.io/posts/git-commit/
Because there are so many ineffective commit messages in the PR, please use "Squash and Merge" when this PR is eventually approved and it is time to merge.

Please address the changes above and those noted in the code, then notify me and I will re-review.

Thanks!

**Comment by mikebaumann** [comment]
As a check to verify current operation, I selected Travis "Restart Build" in for `continuous-integration/travis-ci/push" and it failed with the error that follows, as shown here:
https://travis-ci.org/BD2KGenomics/dcc-ops/builds/370765124#L4677

```
Pulling dcc-action (quay.io/ucsc_cgl/action-service:1.0.12)...
1.0.12: Pulling from ucsc_cgl/action-service
failed to register layer: Untar error on re-exec cmd: fork/exec /proc/self/exe: cannot allocate memory
The command "sudo bash install_bootstrap -t" exited with 1.
16.05s$ sudo docker stop $(sudo docker ps -aq)
18e992e5049b
5d4a70d79d30
208f0f453261
74eae7899a14
4bc7c56dd674
824209116b61
5421e5599a9f
7ebe374ddd05
77e07f26ca53
0e3eaafbbdb9
6a1abaf14589
17fc9c38a0f9
f9c83ee6ae7e
3d7744aefb65
e0f326d75f9d
8533d311d590
2ee4cc421350
d6f73ffb9749
5c513adfc40e
5651541543fb
678e9992f6b0
75b35680ab67
dd1bcd7c6e68
f9147b6e4014
95beb5da96dd
ff2b7dadc792
12ca6d392440
The command "sudo docker stop $(sudo docker ps -aq)" exited with 0.
Done. Your build exited with 1.
```

**Comment by wshands** [comment]
Closing this PR since dcc-ops is no longer worked on

**Pull request closed without merging**